File: app/src/main/java/me/magnum/melonds/MelonTouchHandler.java
package me.magnum.melonds;

import me.magnum.melonds.model.Input;
import me.magnum.melonds.model.Point;

public class MelonTouchHandler implements IInputListener {
	@Override
	public void onKeyPress(Input key) {
		MelonEmulator.onInputDown(key);
	}

	@Override
	public void onKeyReleased(Input key) {
		MelonEmulator.onInputUp(key);
	}

	@Override
	public void onTouch(Point point) {
		MelonEmulator.onScreenTouch(point.x, point.y);
	}
}

File: app/src/main/java/me/magnum/melonds/ui/emulator/input/MelonTouchHandler.kt
package me.magnum.melonds.ui.emulator.input

import me.magnum.melonds.MelonEmulator
import me.magnum.melonds.domain.model.Input
import me.magnum.melonds.domain.model.Point

class MelonTouchHandler : IInputListener {
    private var isLidClosed = false

    override fun onKeyPress(key: Input) {
        if (key == Input.HINGE) {
            handleHingePress()
        } else {
            MelonEmulator.onInputDown(key)
        }
    }

    override fun onKeyReleased(key: Input) {
        if (key != Input.HINGE) {
            MelonEmulator.onInputUp(key)
        }
    }

    override fun onTouch(point: Point) {
        MelonEmulator.onScreenTouch(point.x, point.y)
    }

    fun setLidClosed(closed: Boolean) {
        if (closed != isLidClosed) {
            isLidClosed = closed
            if (isLidClosed) {
                MelonEmulator.onInputDown(Input.HINGE)
            } else {
                MelonEmulator.onInputUp(Input.HINGE)
            }
        }
    }

    private fun handleHingePress() {
        setLidClosed(!isLidClosed)
    }
}